kernel/src/hil/asymmetric_encryption.rs

```diff
@@ -0,0 +1,133 @@
+    /// cannot be used and will return `None` until the upcall has been called.
+    ///
+    /// The keys generated by `generate()` will depend on the implementation.
+    fn generate(&'a self, trng: &'a dyn entropy::Entropy32, length: usize)
+        -> Result<(), ErrorCode>;
+
+    /// Import an existing public key. This will override any existing keys.
+    fn import_public_key(&mut self, public_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Import an existing private key. This will override any existing keys.
+    fn import_private_key(&mut self, private_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Store the public key in buffer if it exists.
+    fn pub_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+
+    /// Store the private key in buffer if it exists.
+    fn priv_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+}
+
+pub enum Operation {
+    Encrypt,
+    Decrypt,
+    Sign,
+    Verify,
+}
+
+/// Upcall from the `RsaCrypto` trait.
+pub trait Client<'a, B> {
+    /// The `encrypt()`/`decrypt()`/`sign()` command has been completed.
+    ///
+    /// On success the value of `result` should be ignored unless the `op`
+    /// was a `Operation::Verify` in which cast this indicates if the message
+    /// is valid.
+    fn operation_done(&'a self, result: Result<bool, ErrorCode>, op: Operation, buffers: B);
+}
+
+/// The RSA specific crypto buffers
+/// encrypt -> Source, destination and key.
+/// verify -> Message, signature and key.
+/// decrypt -> Source, destination and key.
+/// Sign -> Input, destination and key.
+pub struct RsaCryptoBuffers {
+    /// `encrypt()` -> Source
+    /// `verify()` -> Message
+    /// `decrypt()` -> Source
+    /// `sign()` -> Input
+    pub source: &'static mut [u8],
+    /// `encrypt()` -> Destination
+    /// `verify()` -> Signature
+    /// `decrypt()` -> Destination
+    /// `sign()` -> Destination
+    pub dest: &'static mut [u8],
+    pub key: &'static mut dyn RsaKey<'static>,
+}
+
+pub trait RsaKey<'a>: PubPrivKey<'a> {
+    /// Returns the modulus of the key pair if it exists
+    fn modulus(&'a self) -> Option<&'a [u8]>;
+
+    /// Returns the public exponent of the key pair if it exists
+    fn public_exponent(&'a self) -> Option<u32>;
+
+    /// Returns the private exponent of the key pair if it exists
+    fn private_exponent(&'a self) -> Option<&'a [u8]>;
+}
+
+pub trait RsaCrypto<'a> {
+    /// Set the `Client` client to be called on completion.
+    fn set_client(&'a self, client: &'a dyn Client<'a, RsaCryptoBuffers>);
+
+    /// Encrypt the data in the `source` destination with the public `key`
+    /// specified and store the data in the `dest` buffer.
+    ///
+    /// On completion the `operation_done()` upcall will be scheduled.
+    /// The type of encryption that occurs will be based on the latest
+    /// `set_mode*()` function that was called.
+    fn encrypt(&self, buffers: RsaCryptoBuffers) -> Result<(), (ErrorCode, RsaCryptoBuffers)>;
```

Comment:
Since you are moving `RsaCryptoBuffers`,  why not just have parameters?

Reply:
Looking at the capsule code using this I think the `RsaCryptoBuffers` is easier to understand. We pass the buffers in and get them back for every command